CIP-27: Order Quote Rewards

bh2smith opened the pull request:
This PR implements quote rewards as specified in [CIP-27](https://snapshot.org/#/cow.eth/proposal/0x64e061568e86e8d2eec344d4a892e4126172b992cabe59a0b24c51c4c7e6cc33).

- The query `order_quotes.sql` returns the number of quotes a solver provided for executed market orders.
- This number is multiplied with the `QUOTE_REWARD` which is 9 COW tokens. It becomes an additional independent column in the payment data frame `merged_df`.
- Additional COW transfers are generated for solvers who earned quote rewards (even if the solver produced an overdraft for batch rewards).

bh2smith committed:
update web3 and safe dependencies + pin sqlalchemy version

bh2smith committed:
create oreder quote query

bh2smith committed:
start adding quote reward logic

harisang commented on queries/orderbook/order_quotes.sql:
```diff
@@ -0,0 +1,9 @@
+with winning_quotes as (SELECT concat('0x', encode(oq.solver, 'hex')) as solver,
```
Not claiming anything is wrong with that, but somehow I would feel safer to join also with the `orders` table on `uid`, and ask explicitly that the `class` of the order is `market`.

Will need to think if limit orders create any issues with the accounting here. (it really depends on what the `order_quotes` table contains for such orders).

As a reminder, CIP-27 specifies that it only cares about market orders.

bh2smith replied:
I have addressed this comment (visible a couple of lines below).

harisang commented on src/fetch/payouts.py:
```diff
@@ -344,10 +344,16 @@ def post_cip20_payouts(
 
+    quote_rewards_df = orderbook.get_quote_rewards(dune.start_block, dune.end_block)
+    batch_rewards_df = orderbook.get_solver_rewards(dune.start_block, dune.end_block)
+    merged_df = pandas.merge(
```
If I understand correctly, this means that the two payouts (solver and quote) remain completely disjoint, right? I think this is the cleanest approach, and the only "bad" thing about this is that it would lead to doing twice as many COW transfers as we are doing now. But I think it's fine, since verification will remain straightforward, and in any case even now we are a bit wasteful in our transfers.

fhenneke replied:
That logic is not implemented and it is up to us to decide how we want to use the data. We could merge payments if we wanted. You seem to prefer separate payments?
The main advantage of merging other payments (slippage, performance rewards, consistency rewards) was fewer overdrafts and less transfers. Some disadvantages are e.g. that payments are not independent and an error in one of them makes the total payment wrong, or that the logic for computing payments becomes harder to check.

bh2smith replied:
Note that we will need to ensure we are doing a _full outer join_ here and filling null values with zeroes something like `pandas.fillna(0)`.

This would make up for scenarios where solvers are solving, but not quoting and vice versa. Does CIP-27 say anything about solvers to only give quotes and do not solve? Because they do not even need to be registered solvers to do this.

harisang replied:
Per CIP-27, all estimators need to be solvers, and the estimation rewards will be sent to the corresponding solver rewards addresses.

However, it can indeed happen that a solver solves nothing but successfully quotes a few orders.

In any case, I would strongly suggest to keep things completely separate so that debugging/verifying is easier.

bh2smith replied:
It appears this is what has been done, so I will resolve this issue.

bh2smith committed:
keep only market orders

bh2smith commented on .github/workflows/postgres.yaml:
```diff
@@ -26,6 +26,7 @@ jobs:
         uses: actions/setup-python@v3
         with:
           python-version: '3.10'
+
```
```suggestion
```

bh2smith commented on queries/orderbook/order_quotes.sql:
```diff
@@ -0,0 +1,12 @@
+
+SELECT solver, count(*) AS num_quotes
+FROM winning_quotes
+GROUP BY solver
```
```suggestion
GROUP BY solver

```

bh2smith committed:
Apply suggestions from code review

harisang commented:
This looks good! Haven't tested anything yet though but it makes sense. Is there anything still missing? Asking since it is still marked as Draft.

bh2smith commented:
>  Is there anything still missing?

1. there is still logic missing that connects all this to the payout construction (right now it is just the acquisition of relevant data)
2. Should probably also test this method.
3. Still need to reduce the `COW_BUDGET`

I thought that @fhenneke could branch off of this and implement these items, then they could be merged here and this could all be merged at once when we are ready to "go live". 

If that's the approach, I would prefer to leave this branch in draft state until everything has come together.

fhenneke committed:
added quote rewards payment logic

- changed batch budget
- added quote_reward_cow column to data frames
- added additional transfer for quote rewards

fhenneke commented:
I have read the CLA Document and I hereby sign the CLA

fhenneke commented:
I added some logic around creating transfers for quote rewards: If the query returns a non-zero amount of reward-worthy quotes for a solver, an additional transfer of COW tokens is created.

I was not able to get the code running locally, so this is not tested. (My python environment seems be configured incorrectly.)

bh2smith commented on src/fetch/payouts.py:
```diff
@@ -153,6 +160,21 @@ def as_payouts(self) -> list[Transfer]:
+            except AssertionError:
+                logging.warning(
+                    f"Invalid COW Transfer {self.solver} with amount={quote_reward_cow}"
+                )
```
I don't think this is necessary here. If I remember correctly, we have it in other places because the `amount_wei` could be non-positive.

fhenneke replied:
Yes, they are not necessary. I will remove ~them~ it.

fhenneke replied:
Should all the "try-catch" blocks be turned into if statements?
Only one of the 5 try statements can be removed without replacement.

bh2smith commented on src/fetch/transfer_file.py:
```diff
@@ -22,7 +22,7 @@
     DOCS_URL,
 )
-from src.fetch.payouts import post_cip20_payouts
+from src.fetch.payouts import post_cip27_payouts
```
In the near future I plan to factor out all pre-cip20 logic from this project and then we will not need this special method name anymore. Thanks for reminding me.

bh2smith replied:
In fact, if we rebased this branch on #311, we would not have to touch this code for a rename.

bh2smith reviewed (commented):
Everything looks pretty good in this last commit, but I am not sure I understand the purpose of all these "try-catch" blocks with assertion errors. Could you elaborate?

bh2smith committed:
updating tests according to changes

bh2smith committed:
cleaning up and trying to fix tests

bh2smith committed:
cow to reward target

fhenneke committed:
changed overdraft logic

bh2smith committed:
fix broken test

bh2smith committed:
lint n shit

bh2smith committed:
[CIP-27] Period Budget (#312)

Easy Update Period Budget and adjust tests accordingly.

bh2smith committed:
fix bad rebase

bh2smith committed:
Merge branch 'main' into cip-27

harisang commented on src/fetch/payouts.py:
```diff
@@ -139,8 +145,19 @@ def as_payouts(self) -> list[Transfer]:
+                    amount_wei=quote_reward_cow,
+                )
+            )
+        if self.is_overdraft():
```
Is there a reason we changed from `assert not self.is)overdraft()` to this if-statement? (sorry for my ignorance here). I guess we are now explicitly returning here the empty list?

fhenneke replied:
Before this change, only rows without an over draft called this function. To make sure, that this is true, there was an assert statement.

After this change, the function is also called in case of overdrafts (as in an overdraft for batch rewards) since quote rewards are independent of batch rewards and this function handles the creation of all transfers to solvers.

If there is an overdraft, only transfers for quote rewards are created and the rest of the function is skipped. This is my there is this if statement.
If there is no overdraft, the rest of the function is executed as normal.

harisang replied:
Oh good, thanks for clarifying!

harisang commented:
Looks good!

bh2smith commented:
So -- we can merge this now, but this code will not run as is until the following Dune Spellbook PR lands:

https://github.com/duneanalytics/spellbook/pull/3713

In the event that this does not land early enough we can circumvent the issue by querying the "raw_order_rewards" table directly.

bh2smith commented:
Update -- that Spellbook PR has been merged. Will merge this too -- but not going to tag the release yet so that next week is still pre-cip27
